### PR TITLE
bindinfo: support 'show binding_cache status'

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/32466

Problem Summary:
Support 'show binding_cache status' to show the information about the bindCache.

### What is changed and how it works?
1. Modify the parser to support the grammar for `show binding_cache status`
2. Show the "num_bindings", "total_bindings", "memory_usage", "memory_capacity" for `show binding_cache status`

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Reminiscent** [comment]
wait for https://github.com/pingcap/tidb/pull/32439 merge

File: executor/show.go [hunk 1]
```diff
@@ -340,6 +342,38 @@ func (e *ShowExec) fetchShowBind() error {
 	return nil
 }
 
+func (e *ShowExec) fetchShowBindingCacheStatus(ctx context.Context) error {
+	// TODO: Should we need to check the privilege for the mysql.bind_info table
+	exec := e.ctx.(sqlexec.RestrictedSQLExecutor)
+
+	rows, _, err := exec.ExecRestrictedSQL(ctx, nil, `SELECT count(*) FROM mysql.bind_info where status = 'using';`)
```
Comment (qw4990) [inline executor/show.go:349]: Should we use `where status in ('using', 'enable')` here?

File: executor/show.go [hunk 2]
```diff
@@ -340,6 +342,38 @@ func (e *ShowExec) fetchShowBind() error {
...
+		return errors.Trace(err)
+	}
+
+	handle := domain.GetDomain(e.ctx).BindHandle()
+
+	bindRecords := handle.GetAllBindRecord()
+	numBindings := 0
+	for _, bindRecord := range bindRecords {
+		for _, binding := range bindRecord.Bindings {
+			if binding.Status == bindinfo.Using {
```
Comment (Reminiscent) [inline executor/show.go:360]: Wait for https://github.com/pingcap/tidb/pull/32801

File: executor/show.go [hunk 3]
```diff
@@ -340,6 +342,37 @@ func (e *ShowExec) fetchShowBind() error {
 	return nil
 }
 
+func (e *ShowExec) fetchShowBindingCacheStatus(ctx context.Context) error {
+	exec := e.ctx.(sqlexec.RestrictedSQLExecutor)
+
+	rows, _, err := exec.ExecRestrictedSQL(ctx, nil, `SELECT count(*) FROM mysql.bind_info where status = 'using' or status = 'enabled';`)
```
Comment (qw4990) [inline executor/show.go:348]: How about using constants `bindinfo.Enabled/Using` instead of string literals.

File: executor/show_test.go [hunk 4]
```diff
@@ -1759,3 +1759,38 @@ func TestShowBindingCache(t *testing.T) {
...
 }
+
+func TestShowBindingCacheStatus(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustQuery("show binding_cache status").Check(testkit.Rows(
+		"0 0 0 67108864"))
```
Comment (qw4990) [inline executor/show_test.go:1770]: The byte numbers seems not readable, how about using `util/memory.FormatBytes` to format them?

File: executor/show_test.go [hunk 5]
```diff
@@ -1759,3 +1759,38 @@ func TestShowBindingCache(t *testing.T) {
...
+	rows := result.Rows()
+	require.Equal(t, len(rows), 0)
+	tk.MustExec("create global binding for select * from t using select * from t")
+
+	result = tk.MustQuery("show global bindings")
+	rows = result.Rows()
+	require.Equal(t, len(rows), 1)
+
+	tk.MustQuery("show binding_cache status").Check(testkit.Rows(
+		"1 1 159 67108864"))
```
Comment (qw4990) [inline executor/show_test.go:1784]: Please delete/remove this binding and then show its status and check the result again.

**Review by qw4990 — APPROVED** [review]

File: executor/show.go [hunk 6]
```diff
@@ -340,6 +343,37 @@ func (e *ShowExec) fetchShowBind() error {
 	return nil
 }
 
+func (e *ShowExec) fetchShowBindingCacheStatus(ctx context.Context) error {
+	exec := e.ctx.(sqlexec.RestrictedSQLExecutor)
+
+	rows, _, err := exec.ExecRestrictedSQL(ctx, nil, fmt.Sprintf("SELECT count(*) FROM mysql.bind_info where status = '%s' or status = '%s';", bindinfo.Enabled, bindinfo.Using))
```
Comment (Reminiscent) [inline executor/show.go:349]: Need to add the `Disabled` status binding.

Comment (Reminiscent) [inline executor/show.go:349]: Need to add the `Disabled` status binding.

File: planner/core/planbuilder.go [hunk 7]
```diff
@@ -4714,6 +4714,9 @@ func buildShowSchema(s *ast.ShowStmt, isView bool, isSequence bool) (schema *exp
 	case ast.ShowBindings:
 		names = []string{"Original_sql", "Bind_sql", "Default_db", "Status", "Create_time", "Update_time", "Charset", "Collation", "Source"}
 		ftypes = []byte{mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeDatetime, mysql.TypeDatetime, mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeVarchar}
+	case ast.ShowBindingCacheStatus:
+		names = []string{"num_bindings", "total_bindings", "memory_usage", "memory_capacity"}
```
Comment (time-and-fate) [inline planner/core/planbuilder.go:4718]: I think the difference between the first two columns is not so straightforward. How about name them something like "bindings_in_cache" and "bindings_in_table"?
And I think "memory_capacity" can be named "memory_quota" to be consistent with the config variable.

File: executor/show.go (hunk 6 above)
Comment (Reminiscent) [reply executor/show.go:348]: The tests for `Dsiabled= status` will be added until the PR https://github.com/pingcap/tidb/pull/32819 is merged.

File: executor/show.go [hunk 8]
```diff
@@ -339,6 +342,37 @@ func (e *ShowExec) fetchShowBind() error {
 	return nil
 }
 
+func (e *ShowExec) fetchShowBindingCacheStatus(ctx context.Context) error {
+	exec := e.ctx.(sqlexec.RestrictedSQLExecutor)
+
+	rows, _, err := exec.ExecRestrictedSQL(ctx, nil, fmt.Sprintf("SELECT count(*) FROM mysql.bind_info where status = '%s' or status = '%s' or status = '%s';", bindinfo.Enabled, bindinfo.Using, bindinfo.Disabled))
```
Comment (time-and-fate) [inline executor/show.go:348]: So is it possible that bindings on a single SQL are counted multiple times?
Seems the the logic below only counts the enabled bindings.

Comment (Reminiscent) [reply executor/show.go]: No, a single SQL can only contain one binding at most. Change it to only count the enabled binding.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**